docs/architecture-decisions/2024-01-13_native_binaries.md

```diff
@@ -0,0 +1,182 @@
+
+#### PoC
+
+The PoC is available [here](https://github.com/algorandfoundation/algokit-cli/pull/393). It outlines a simple github action with extra setup that compiles algokit cli as a single file executable on latest versions of Windows, Mac and Linux github runners.
+
+### Option 3 - Cross platform python wheels with `cibuildwheel`
+
+**Pros**
+
+- cibuildwheel is a versatile tool that simplifies the process of building wheels for different platforms and Python versions.
+- It supports a wide range of Python versions and platforms, including Windows, macOS, and Linux.
+- It integrates well with CI/CD systems, making it easy to automate the wheel building process.
+- It can handle complex dependencies and package structures.
+- It provides a consistent and reproducible build environment, which can help reduce environment-specific bugs.
+- Minimal configuration is required to get started and aligns with existing distribution methods like PyPi.
+
+**Cons**
+
+- As it doesn't produce standalone executables, the end users still need to have Python installed on their systems.
+- The wheels produced might not be as portable as standalone executables, as they can still be affected by the specific Python environment on the user's machine.
+- It doesn't inherently support distribution methods like Homebrew or Winget, additional work might be needed to integrate with these systems.
+
+There is no PoC supporting this option given an official github action by `pypa` that is a plug and play solution for building wheels for different platforms and python versions.
+
+### Benchmarking `pyinstaller` vs `nuitka` vs pipx installed `algokit`
+
+#### Methodology
+
+`hyperfine` was used to benchmark 5 different executables:
+
+- Nuitka Onefile - Nuitka compiled executable with `--onefile` flag, which produces a single file executable.
+- Nuitka Onedir - Nuitka compiled executable with `--onedir` flag, which produces a directory with the executable and other dependencies unzipped.
+- PyInstaller Onedir - PyInstaller compiled executable with `--standalone` flag, which produces a directory with the executable and other dependencies unzipped.
+- PyInstaller Onefile - PyInstaller compiled executable with `--onefile` flag, which produces a single file executable.
+- AlgoKit from `pipx` - AlgoKit CLI installed via `pipx` with all dependencies frozen (current latest stable release).
+
+The benchmarking was performed on a MacBook M2 running macOS 14.2.1 and an ARM based Ubuntu 20.04.3 LTS running on a Parallels Desktop on the same machine.
+
+#### Results
+
+| Method              | macOS M2 | Ubuntu 20 ARM Linux VM | Windows 11 ARM |
+| ------------------- | -------- | ---------------------- | -------------- |
+| nuitka_onefile      | 3.634    | 1.465                  | 3.874          |
+| nuitka_onedir       | 0.2515   | 0.6200                 | 0.5136         |
+| pyinstaller_onedir  | 0.3228   | 0.7927                 | 0.6668         |
```

Comment:
Given simplicity and lower build time, this performance is pretty on par with nuitka_onedir?

Reply:
@robdmoore yes id say nuitka_onedir, pyinstaller_onedir and algokit via pypi is within a margin of milliseconds - still figured its worth outlining that nuitka indeed performs fastest given that its precompiled into C. From usage perspective though pyinstaller_onedir is comparably fast so wouldn't consider nuitka a huge advantage in terms of speed